Add name, description, and credits fields to editor screen

JoshuaSchaefer opened the pull request:
Moved the name, description, and credit editors into the project editor screen. Users can now view and update all the fields at the same time.
---
### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroweb-Symfony/blob/develop/.github/contributing.md) and [wiki pages](https://github.com/Catrobat/catroweb-Symfony/wiki/) of this repository.

- [ ] Include the name and id of the Jira ticket in the PR’s title eg.: `SHARE-666 The devils ticket`
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no warnings and errors 
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Verify that all tests are passing (CI), if not please state the test cases in the [section](#Tests) below
- [ ] Perform a self-review of the changes
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] Put your ticket into the `Code Review` section in [Jira](https://jira.catrob.at/)
- [ ] Post a message in the *#catroweb* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer
- [ ] Check that your pull request has been successfully deployed to https://web-test-1.catrob.at/

### Additional Description
This allows users to update their default and custom translations for the name, description and credits of their project on one full screen.

![Screen Shot 2022-02-15 at 11 44 51 PM](https://user-images.githubusercontent.com/38061152/154197794-455ea31d-1e54-4ac1-bcb2-ee74e5cad920.png)


### Tests - additional information
Updated existing tests to reflect changes and created new tests to unify redundant existing tests for the editor.

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+      return
+    }
+
+    if (!this.hasDefaultText && !(languageSelected === '' || languageSelected === 'default')) {
```
this condition is a bit confusing. maybe refactor to a function with an explanatory name or put this as an` else if` block after the next `if` block?

JoshuaSchaefer replied:
Refactored this section a bit to simplify the logic

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+    const newText = this.editText.val().trim()
+    if (this.areChangesSaved()) {
+      this.close()
+      return
```
i'm not sure how js handles the return type of `this.save`. the main return clauses return `Promise` but this returns nothing. How does it behave when `ProjectEditor` calls this function?

JoshuaSchaefer replied:
I'm not 100% sure how this works, but from my testing, it looks like it returns as if a promise is resolved if you don't return anything. So, it practically acts like a successful save when it returns

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  this.getNewText = (language) => {
+    if (language === 'default') {
```
LOL this is a bug fix? if so, is there a new test for this? I didn't check

JoshuaSchaefer replied:
This isn't a bug fix. Before if there wasn't anything defined for the text field, then the language selector didn't show. However, we need to show it now because they could have something defined for description and not credits. So, now the text field is disabled if there is not a default text defined and there is a different language selected if that makes sense. For example, if a default description is defined, but no default credits. When the language selected is french, then the description will be enabled and the credit will be disabled.

I will add a test to make sure it is disabled

danielzhang130 replied:
IC. Do we really need that complexity of enabling certain fields and not others? I think it's fine to just let the user define whatever they want.

Also i think there was a bug. 
![image](https://user-images.githubusercontent.com/37479705/154365956-eb19a6da-bd9b-4353-a7e0-84cdfed300d9.png)
That's what happens if you try to edit the title, change to a language then change back to default.

JoshuaSchaefer replied:
Previously, you couldn't define a translation without defining a default. So, I was just trying to continue the previous logic by disabling the fields that don't have a default. I don't really think it makes sense to be able to define a translation without a default as well.

Also, I just ran that test case and I didn't get that API call. What did you click when the keep or discard text prompt came up?

danielzhang130 replied:
when i did it, i didn't edit anything, i just went to the edit page and directly changed the language, then change back to default

JoshuaSchaefer replied:
Oh, okay. I just ran that test as well and I still didn't get that API call. Is that on your local or on prod?

JoshuaSchaefer replied:
Bug should be fixed in this PR

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+  this.pathEditDescription = projectDescriptionCredits.data('path-edit-program-description')
+  this.pathEditCredits = projectDescriptionCredits.data('path-edit-program-credits')
+
+  this.show = () => {
```
When i think of it, this `ProjectEditorTextField` is never independently shown or hidden. it is shown when the `ProjectEditor` is shown. So I think this function is more like a `setText`

JoshuaSchaefer replied:
Updated to setText

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -0,0 +1,202 @@
+      langaugeSelectElement.addClass('d-none')
+    }
+
+    window.history.pushState(
```
i forgot to ask in that pr. Is this for the browser "navigate back" button, or something else?

JoshuaSchaefer replied:
Yeah, it practically allows you to return to the main program screen by clicking the back button

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -0,0 +1,202 @@
+    )
+
+    $(window).on('popstate', this.popStateHandler)
+    showCustomTopBarTitle('Program Editor', function () {
```
Translatable text. Also is "Program Editor" a good name? what if we just leave the title empty?

JoshuaSchaefer replied:
I'm good just to leave it empty

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -0,0 +1,202 @@
+
+  $(document).ready(getLanguages)
+
+  editProgramButton.on('click', () => {
```
i feel like this should be somewhere outside of this class. so the public interface of this class should just be `show()`

JoshuaSchaefer replied:
Moved to the ProgramName function

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+    this.lastSavedText = this.initialText
+  }
+
+  this.close = () => {
```
This class also doesn't have a `close` action logically. This is more like a `hideError`

JoshuaSchaefer replied:
Updated to hideError

danielzhang130 replied:
Can we move this next to `showError`?

JoshuaSchaefer replied:
Moved next to showError

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  function saveCustomTranslationSuccess (language) {
+    self.snackbar.show(self.translationSaved, language)
```
so this screen will have 3 snackbars? I think we could use only one.

JoshuaSchaefer replied:
Refactored this logic to only show one snackbar for save and one for delete

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,163 @@
+    )
+  }
+
+  this.keepChanges = () => {
```
`keepChanges` to me sound like a noop function. Also is it a bug if I do:
change text, then switch language, then select keep changes in the dialog, then press back button, then no "save changes" dialog will be shown?
I'm not sure if this happens, I haven't tried.

JoshuaSchaefer replied:
Updated this keepChanges section to get the saved custom translation or default text as the lastSavedText. Also added some logic to disable or enable text fields

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,184 @@
+  this.pathEditDescription = projectDescriptionCredits.data('path-edit-program-description')
+  this.pathEditCredits = projectDescriptionCredits.data('path-edit-program-credits')
+
+  this.setText = () => {
```
Since `initialText` is never reassigned, why can't the content of this `setText` function be executed in the constructor of the class?

JoshuaSchaefer replied:
the setText is called each time the editor is opened because it should be set to the default text.

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,184 @@
+  this.textLoadingSpinner = $('#edit-' + this.programSection + '-loading-spinner')
+  this.selectedLanguage = $('#edit-selected-language')
+
+  this.defaultSection = $('#' + this.programSection)
```
Is this only used by the next line?

JoshuaSchaefer replied:
Yup, moved it to the next line

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,184 @@
+
+  this.save = (languageSelected) => {
+    const newText = this.editText.val().trim()
+    if (this.areChangesSaved() || this.isDisabled(languageSelected)) {
```
I'm still quite confused about why we should `hideError` upon this specific condition. Can't we just `hideError` as a first thing in the `save` functions? Isn't that simpler and still work?

JoshuaSchaefer replied:
Moved hideError to the beginning of this function

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,184 @@
+      this.editText.removeAttr('disabled')
+      this.lastSavedText = this.initialText
+    } else {
+      this.customTranslationApi.getCustomTranslation(
```
should we show a loading spinner here?

JoshuaSchaefer replied:
I think it is fine if this loads in the background. They would have to save it pretty fast for this to cause a problem

danielzhang130 replied:
ok

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,184 @@
+    return this.editText.val().trim() === this.lastSavedText.trim()
+  }
+
+  this.isDisabled = (languageSelected) => {
```
I think this is more like `shouldDisable` rather than `isDisabled`
I think it might be appropriate here to also include a comment on the design decision of disabling the text field if the field is not defined in the default language

JoshuaSchaefer replied:
Renamed to shouldDisable and added a comment explaining the logic

danielzhang130 commented on tests/BehatFeatures/web/project-details/project_editor.feature:
```diff
@@ -0,0 +1,68 @@
+      | 1  | Catrobat  |
+      | 2  | OtherUser |
+    And there are projects:
+      | id | name      | owned by  | description   | credit
```
```suggestion
      | id | name      | owned by  | description   | credit |
```

danielzhang130 commented on tests/BehatFeatures/web/project-details/project_editor.feature:
```diff
@@ -0,0 +1,68 @@
+    Then the element "#edit-text-ui" should be visible
+    And the element "#edit-submit-button" should be visible
+    When I fill in "edit-name-text" with "This is a new name"
+    When I fill in "edit-description-text" with "This is a new description"
```
```suggestion
    And I fill in "edit-description-text" with "This is a new description"
```

danielzhang130 commented on tests/BehatFeatures/web/project-details/project_editor.feature:
```diff
@@ -0,0 +1,68 @@
+    And the element "#edit-submit-button" should be visible
+    When I fill in "edit-name-text" with "This is a new name"
+    When I fill in "edit-description-text" with "This is a new description"
+    When I fill in "edit-credits-text" with "This is a new credit"
```
```suggestion
    And I fill in "edit-credits-text" with "This is a new credit"
```

JoshuaSchaefer commented on assets/js/components/ProjectEditor.js:
```diff
@@ -0,0 +1,245 @@
+      } else if (i < sections.length - 1) {
+        sectionText += ', ' + sections[i] + ','
+      } else {
+        sectionText += ' and ' + sections[i]
```
Need to make translatable

danielzhang130 replied:
hold on. it doesn't have to include the sections. it could just say "changes saved" or like "changes for <language> saved"

JoshuaSchaefer replied:
Okay, so in the case where a user deletes the translation for the description and updates the translation for the credits, there should be a notification saying translation saved and another one saying translation deleted? Or just one notification saying translation updated?

danielzhang130 replied:
I think one snackbar is enough, since everything is successful, there's no more action required from the user. plus these two snackbars are going to be shown on top of each other?

JoshuaSchaefer replied:
Okay, combined to one snackbar that says that the translation is updated.

Also, from my testing, it looks like the snackbars show one after the other. Might be a timing thing though so they could potentially show at the same time.

danielzhang130 commented on tests/BehatFeatures/web/translation/description_custom_translation.feature:
```diff
@@ -9,87 +9,62 @@ Feature: Projects should have descriptions where a custom translation can be def
-    And the element "#edit-language-selector" should be visible
-    And the element "#edit-submit-button" should be visible
-
   Scenario: Adding a custom description translation
```
I think this test case can also be deleted

JoshuaSchaefer replied:
Deleted the test case

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  this.shouldDisable = (languageSelected) => {
+    // Should disable text field if there is not any default text
+    // and language selected is not default
```
```suggestion
    // User should not be able to define custom translation if field is not defined in default language
```

danielzhang130 commented:
Can we order the functions in `ProjectEditor` and `ProjectEditorTextField` by public and private usage?

JoshuaSchaefer commented:
> Can we order the functions in `ProjectEditor` and `ProjectEditorTextField` by public and private usage?

Added private regions in both of these files

danielzhang130 commented:
rebase to get the fix for flaky tests

danielzhang130 commented on assets/js/project.js:
```diff
@@ -29,13 +29,37 @@ const $appLanguage = $('#app-language')
+  const creditsEditorTextField = new ProjectEditorTextField(
+    $projectDescriptionCredits,
+    $projectDescriptionCredits.data('project-id'),
+    'credits',
```
This should be like this for the api to work
```suggestion
    'credit',
```

danielzhang130 commented on tests/BehatFeatures/web/translation/credits_custom_translation.feature:
```diff
@@ -8,96 +8,94 @@ Feature: Projects should have credits where a custom translation can be defined
     And there are projects:
       | id | name      | owned by | credit     |
       | 1  | project 1 | Catrobat | my credits |
-    And I wait 1000 milliseconds
```
This line should not be deleted

JoshuaSchaefer replied:
Added back

danielzhang130 commented on tests/BehatFeatures/web/translation/credits_custom_translation.feature:
```diff
@@ -8,96 +8,97 @@ Feature: Projects should have credits where a custom translation can be defined
-    And the element "#edit-language-selector" should be visible
-    And the element "#edit-submit-button" should be visible
-
   Scenario: Adding a custom credits translation
```
this and the one in name_custom_translation can be delelted to keep consistent with description_custom_translation

JoshuaSchaefer replied:
Deleted

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -0,0 +1,233 @@
+    }
+  }
+
+  this.getSectionText = (sections) => {
```
unused

JoshuaSchaefer replied:
Removed

danielzhang130 commented on assets/js/components/ProjectEditorTextField.js:
```diff
@@ -0,0 +1,175 @@
+    )
+  }
+
+  this.keepChanges = (language) => {
```
this name is very confusing. the function is very similar to getNewText, except it doesn't touch the entered text in the editor. so maybe a name like getNewTextKeepChanges?

JoshuaSchaefer replied:
Renamed

danielzhang130 commented on tests/BehatFeatures/web/translation/name_custom_translation.feature:
```diff
@@ -10,90 +10,65 @@ Feature: Projects should have a name where a custom translation can be defined
-    And the element "#edit-language-selector" should be visible
-    And the element "#edit-submit-button" should be visible
-
   Scenario: Adding a custom name translation
```
this scenario is no longer needed

JoshuaSchaefer replied:
Removed

danielzhang130 reviewed (approved):
almost done

Commit:
Add name, description, and credits fields to editor screen

Moved the name, description, and credit editors into the project editor screen. Users can now view and update all the fields at the same time.
